lld/ELF/InputFiles.cpp

```diff
@@ -1418,6 +1423,67 @@ std::vector<uint32_t> SharedFile::parseVerneed(const ELFFile<ELFT> &obj,
   return verneeds;
 }
 
+// To determine if a shared file can support any of the GNU Attributes,
+// the .note.gnu.properties section need to be read. This has to be done
+// differently for SharedFiles as the information available is not as
+// extensive as a normal object input file. This will take the program
+// headers, along with the SHT_NOTE section header to find the relevant
+// information. This uses a similar process to the readGnuProperty
+// function, but designed specifically for SharedFiles.
+template <typename ELFT>
+void SharedFile::parseGnuAttributes(const uint8_t *base,
+                                    const typename ELFT::PhdrRange headers,
+                                    const typename ELFT::Shdr *sHeader) {
+  auto err = [&](const uint8_t *place) -> ELFSyncStream {
+    auto diag = Err(ctx);
+    diag << this->getName() << ":(" << ".note.gnu.properties" << "+0x"
+         << Twine::utohexstr(place - base) << "): ";
+    return diag;
+  };
+
+  if (numElfPhdrs == 0 || sHeader == nullptr)
+    return;
+  uint32_t featureAndType = ctx.arg.emachine == EM_AARCH64
+                                ? GNU_PROPERTY_AARCH64_FEATURE_1_AND
+                                : GNU_PROPERTY_X86_FEATURE_1_AND;
+
+  for (unsigned i = 0; i < numElfPhdrs; i++) {
+    if (headers[i].p_type != PT_GNU_PROPERTY)
+      continue;
+
+    const typename ELFT::Note note(
+        *reinterpret_cast<const typename ELFT::Nhdr *>(base +
+                                                       headers[i].p_vaddr));
+    if (note.getType() != NT_GNU_PROPERTY_TYPE_0 || note.getName() != "GNU")
+      continue;
+
+    // Read a body of a NOTE record, which consists of type-length-value fields.
```

Comment:
Would it be possible to extract this part so that it is common with readGnuProperty?

I think at this stage both are reading from an `ArrayRef<uint8_t>`